Mathlib/CategoryTheory/MorphismProperty/Comma.lean

```diff
@@ -0,0 +1,244 @@
+where the structural morphism satisfies `P` and where the horizontal morphisms satisfy `Q`.
+Common examples of the latter are e.g. the category of schemes étale (finite, affine, etc.)
+over a base `X`. Here `Q = ⊤`.
+
+## Implementation details
+
+- We provide the general constructor `P.Comma L R Q W` to obtain `Over X` and `Under X` as
+  special cases of the more general setup.
+
+- Most results are developed only in the case where `Q = ⊤` and `W = ⊤`, but the definition
+  is setup in the general case to allow for a later generalization if needed.
+
+-/
+
+namespace CategoryTheory.MorphismProperty
+
+open Limits
+
+section Comma
+
+variable {A : Type*} [Category A] {B : Type*} [Category B] {T : Type*} [Category T]
+  (L : A ⥤ T) (R : B ⥤ T)
+
+variable (P : MorphismProperty T) (Q : MorphismProperty A) (W : MorphismProperty B)
+
+/-- `P.Comma L R Q W` is the subcategory of `Comma L R` consisting of
+objects `X : Comma L R` where `X.hom` satisfies `P`. The morphisms are given by
+morphisms in `Comma L R` where the left one satisfies `Q` and the right one satisfies `W`. -/
+@[ext]
+protected structure Comma (Q : MorphismProperty A) (W : MorphismProperty B) extends Comma L R where
+  prop : P toComma.hom
+
+namespace Comma
+
+variable {L R P Q W}
+
+/-- A morphism in `P.Comma L R Q W` is a morphism in `Comma L R` where the left
+hom satisfies `Q` and the right one satisfies `W`. -/
+@[ext]
+structure Hom (X Y : P.Comma L R Q W) extends CommaMorphism X.toComma Y.toComma where
+  prop_hom_left : Q toCommaMorphism.left
+  prop_hom_right : W toCommaMorphism.right
+
+/-- The underlying morphism of objects in `Comma L R`. -/
+abbrev Hom.hom {X Y : P.Comma L R Q W} (f : Comma.Hom X Y) : X.toComma ⟶ Y.toComma :=
+  f.toCommaMorphism
+
+@[simp, nolint simpVarHead]
+lemma Hom.hom_mk {X Y : P.Comma L R Q W}
+    (f : CommaMorphism X.toComma Y.toComma) (hf) (hg) :
+  Comma.Hom.hom ⟨f, hf, hg⟩ = f := rfl
+
+/-- See Note [custom simps projection] -/
+def Hom.Simps.hom {X Y : P.Comma L R Q W} (f : X.Hom Y) :
+    X.toComma ⟶ Y.toComma :=
+  f.hom
+
+initialize_simps_projections Comma.Hom (toCommaMorphism → hom)
+
+variable [ContainsIdentities Q] [ContainsIdentities W]
+
+/-- The identity morphism of an object in `P.Comma L R Q W`. -/
+@[simps]
+def id (X : P.Comma L R Q W) : Comma.Hom X X where
+  left := 𝟙 X.left
+  prop_hom_left := Q.id_mem X.toComma.left
+  prop_hom_right := W.id_mem X.toComma.right
+
+variable [Q.IsStableUnderComposition] [W.IsStableUnderComposition]
+
+/-- Composition of morphisms in `P.Comma L R Q W`. -/
+@[simps]
+def Hom.comp {X Y Z : P.Comma L R Q W} (f : Comma.Hom X Y) (g : Comma.Hom Y Z) :
+    Comma.Hom X Z where
+  left := f.left ≫ g.left
+  right := f.right ≫ g.right
+  prop_hom_left := Q.comp_mem _ _ f.prop_hom_left g.prop_hom_left
+  prop_hom_right := W.comp_mem _ _ f.prop_hom_right g.prop_hom_right
+
+variable (L R P Q W) in
+instance : Category (P.Comma L R Q W) where
```

Comment:
`ContainsIdentities` or `IsStableUnderComposition` are the sharper assumptions for the definitions `id` and `comp` above, but I would suggest using `IsMultiplicative` as a single assumption for the definition of the category instance.